core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -122,46 +197,65 @@ public void setStoreListenerSource(Cache<K, V> source) {
-        new HashMap<CacheEventNotificationServiceImpl.EventListenerWrapper, Future<?>>(registeredListeners.size());
-    
-    for (final EventListenerWrapper wrapper: registeredListeners) {
-      if (!wrapper.config.fireOn().contains(type)) {
-        continue;
+    CacheEventWrapper<K, V> cacheEventWrapper = new CacheEventWrapper<K, V>(event);
+    if (hasListeners()) {
+      threadLocal.verifyOrderedDispatch(orderedListenerCount.get() != 0);
+      threadLocal.addToEventList(cacheEventWrapper);
+      if (threadLocal.isOrdered()) {
+        ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue(event.getKey());
+        eventWrapperQueue.add(cacheEventWrapper);
       }
-      Runnable notificationTask = new Runnable() {
-        @Override
-        public void run() {
-          CacheEventListener<K, V> listener = wrapper.getListener();
-          listener.onEvent(event);
-        }
-      };
-      
-      ExecutorService eventDelivery = wrapper.config.orderingMode().equals(EventOrdering.UNORDERED) ? unorderedDelivery : orderedDelivery;
-      notificationResults.put(wrapper, eventDelivery.submit(notificationTask));
-    }
-    
-    for (Map.Entry<EventListenerWrapper, Future<?>> entry: notificationResults.entrySet()) {
-      EventListenerWrapper wrapper = entry.getKey();
-      Future<?> f = entry.getValue();
-      if ((EventFiring.SYNCHRONOUS.equals(wrapper.config.firingMode()))) {
-        boolean interrupted = false;
-        try {
-          f.get();
-        } catch (ExecutionException e) {
-            LOGGER.error("Cache Event Listener failed for event type {} due to ", type, e);
-          // XXX delegate to resilience strategy (#52), and/or just log
-        } catch (InterruptedException e) {
-          interrupted = true;
-        } finally {
-          if (interrupted) {
-            Thread.currentThread().interrupt();
-          }
+    }
+  }
+
+  private void dispatchEvent(CacheEventWrapper cacheEventWrapper, EventThreadLocal currentEventThreadLocal) {
+    cacheEventWrapper.markFireable();
+    ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = getEventQueue((K)cacheEventWrapper.cacheEvent.getKey());
+    orderedEventWait(cacheEventWrapper, eventWrapperQueue, currentEventThreadLocal);
+    if (currentEventThreadLocal.isOrdered()) {
+      while (!eventWrapperQueue.isEmpty() && eventWrapperQueue.peek().isFireable()) {
+      CacheEventWrapper event = eventWrapperQueue.remove();
+        actualDispatch(event, orderedEventDispatcher);
+        if (event != currentEventThreadLocal.get().get(0)) {
+          event.markFiredAndSignalCondition(true);
+        } else {
+          event.markFiredAndSignalCondition(false);
         }
       }
+    } else {
+      actualDispatch(cacheEventWrapper, unOrderedEventDispatcher);
+    }
+  }
+
+  private void orderedEventWait(CacheEventWrapper cacheEventWrapper, Queue<CacheEventWrapper> eventWrapperQueue, EventThreadLocal currentEventThreadLocal) {
```

Comment:
Objective – make threads that don’t have event at the front of the queue (but want to fire their event) wait until they are notified hem that their event if FIRED

What it internally does:
•	Check the dispatch is ordered
•	Confirm event is not at the front of the queue
•	Wait for the signal on FIRED_CONDITION, that should be provided by the thread with event at the front

Why it should work:
Having other threads to wait while a single thread fires their events allows for a simple approach to carry out dispatch for all events that are `FIREABLE`. Also since the thread that fires all events signals for all the events that it fires but are not created by that thread makes sure that no thread is kept in a perpetual wait.